tests/returntypes/test_parameter_shift_hessian_new.py

```diff
@@ -0,0 +1,1757 @@
+
+    def test_with_diag_argnum(self):
+        argnum = qml.math.eye(3, dtype=bool)
+        diag, offdiag = _collect_recipes(self.tape, argnum, ("A",) * 3, None, None)
+        for res, exp in zip(diag, self.expected_diag_recipes):
+            assert qml.math.allclose(res, exp)
+        assert all(entry == (None, None, None) for entry in offdiag)
+
+    def test_with_block_diag_argnum(self):
+        argnum = qml.math.array([[True, True, False], [True, True, False], [False, False, True]])
+        diag, offdiag = _collect_recipes(self.tape, argnum, ("A",) * 3, None, None)
+        for res, exp in zip(diag, self.expected_diag_recipes):
+            assert qml.math.allclose(res, exp)
+        assert qml.math.allclose(offdiag[0], self.two_term_recipe)
+        assert qml.math.allclose(offdiag[1], self.four_term_recipe)
+        assert offdiag[2] == (None, None, None)
+
+    def test_with_other_argnum(self):
+        argnum = qml.math.array([[True, True, False], [True, False, True], [False, True, True]])
+        diag, offdiag = _collect_recipes(self.tape, argnum, ("A",) * 3, None, None)
+        for i, (res, exp) in enumerate(zip(diag, self.expected_diag_recipes)):
+            if i == 1:
+                assert res is None
+            else:
+                assert qml.math.allclose(res, exp)
+        assert qml.math.allclose(offdiag[0], self.two_term_recipe)
+        assert qml.math.allclose(offdiag[1], self.four_term_recipe)
+        assert qml.math.allclose(offdiag[2], self.four_term_recipe)
+
+
+class TestGenerateOffDiagTapes:
+    """Test some special features of `_generate_offdiag_tapes`."""
+
+    @pytest.mark.parametrize("add_unshifted", [True, False])
+    def test_with_zero_shifts(self, add_unshifted):
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(np.array(0.2), wires=[0])
+            qml.RY(np.array(0.9), wires=[0])
+
+        recipe_0 = np.array([[-0.5, 1.0, 0.0], [0.5, 1.0, np.pi]])
+        recipe_1 = np.array([[-0.25, 1.0, 0.0], [0.25, 1.0, np.pi]])
+        t, c = [], []
+        new_add_unshifted, unshifted_coeff = _generate_offdiag_tapes(
+            tape, (0, 1), [recipe_0, recipe_1], add_unshifted, t, c
+        )
+
+        assert len(t) == 3 + int(add_unshifted)  # Four tapes of which the first is not shifted
+        assert np.allclose(c, [-0.125, -0.125, 0.125])
+        assert np.isclose(unshifted_coeff, 0.125)
+        assert not new_add_unshifted
+
+        orig_cls = [orig_op.__class__ for orig_op in tape.operations]
+        expected_pars = list(product([0.2, 0.2 + np.pi], [0.9, 0.9 + np.pi]))
+        if not add_unshifted:
+            expected_pars = expected_pars[1:]
+        for exp_par, h_tape in zip(expected_pars, t):
+            assert len(h_tape.operations) == 2
+            assert all(op.__class__ == cls for op, cls in zip(h_tape.operations, orig_cls))
+            assert np.allclose(h_tape.get_parameters(), exp_par)
+
+
+class TestParameterShiftHessian:
```

Comment:
Added these tests for multiple measurements and multiple params, all of them on the tape level.